Sources/TensorFlow/Operators/LinearAlgebra.swift

```diff
@@ -266,3 +266,128 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    } else if matrix.rank > rhs.rank {
+        let leadingDims = extractLeadingDims(rhs.shape, matrix.shape, ignoreLast: 2)
+        let matchedRhs = rhs.broadcasted(to: leadingDims + rhs.shape)
+        return _Raw.matrixTriangularSolve(matrix: matrix, rhs: matchedRhs, lower: lower, adjoint: adjoint)
+    }
+    return _Raw.matrixTriangularSolve(matrix: matrix, rhs: rhs, lower: lower, adjoint: adjoint)
+}
+
+@inlinable
+@derivative(of: triangularSolve)
+internal func _vjpTriangularSolve<T: TensorFlowFloatingPoint>(
+    matrix: Tensor<T>,
+    rhs: Tensor<T>,
+    lower: Bool = true,
+    adjoint: Bool = false
+) -> (value: Tensor<T>, pullback: (Tensor<T>) -> (Tensor<T>, Tensor<T>)) {
+    let broadcastMatrix: Bool = matrix.rank < rhs.rank
+    let broadcastRhs: Bool = matrix.rank > rhs.rank
+
+    var matchedMatrix: Tensor<T> = matrix
+    if broadcastMatrix {
+        let leadingDims = extractLeadingDims(rhs.shape, matrix.shape, ignoreLast: 2)
+        matchedMatrix = matrix.broadcasted(to: leadingDims + matrix.shape)
+    }
+    
+    let x = triangularSolve(matrix: matchedMatrix, rhs: rhs, lower: lower, adjoint: adjoint)
+    let triangularSolveGrad = { (gradient: Tensor<T>) -> (Tensor<T>, Tensor<T>) in
+        let rhsGrad = triangularSolve(matrix: matchedMatrix, rhs: gradient, lower: lower, adjoint: !adjoint)
+        let (left, right) = adjoint ? (x, rhsGrad) : (rhsGrad, x)
+        let matrixGrad = -matmul(left, transposed: false, right, transposed: true)
+        let triMatrixGrad = lower
+            ? matrixGrad.bandPart(subdiagonalCount: -1, superdiagonalCount: 0)
+            : matrixGrad.bandPart(subdiagonalCount: 0, superdiagonalCount: -1)
+        if broadcastMatrix {
+            return (triMatrixGrad.unbroadcasted(to: matrix.shape), rhsGrad)
+        } else if broadcastRhs {
+            return (triMatrixGrad, rhsGrad.unbroadcasted(to: rhs.shape))
+        }
+        return (triMatrixGrad, rhsGrad)
+    }
+    return (x, triangularSolveGrad)
+}
+
+
+/// Returns leading dimension from tensors shapes.
+/// Consider input tensors Left and Right, there are three major cases
+/// - `Left.shape > Right.shape` and `Left.shape` equals `[A, B, C, D, F]`
+///   and `Right.shape` equals `[D, F]`, the `extractLeadingDims` returns `[A, B, C]`
+/// - `Left.shape < Right.shape` and `Left.shape` equals `[D, F]`
+///   and `Right.shape` equals `[A, B, C, D, F]`, the `extractLeadingDims` returns `[A, B, C]`
+/// - `Left.shape == Right.shape`, and `extractLeadingDims` returns `[]`
+///
+/// - Parameters:
+///     - left: Left tensor.
+///     - right: Right tensor.
+///     - ignoreLast: Number of dimensions from the end to ignore. Default value is zero.
+/// - Precondition: Left tensor rank cannot be lower than `ignoreLast`.
+/// - Precondition: Right tensor rank cannot be lower than `ignoreLast`.
+@inlinable
+internal func extractLeadingDims<T: TensorFlowNumeric>(
```

Comment:
`internal` is the default access level and can be omitted.
```suggestion
func extractLeadingDims<T: TensorFlowNumeric>(
```

Reply:
Interesting, why all "func _vjp*" methods use `internal` then?